zebra-network/src/peer/client.rs

```diff
@@ -239,27 +315,108 @@ impl<T: std::fmt::Debug> MustUseOneshotSender<T> {
     }
 }
 
-impl<T: std::fmt::Debug> From<oneshot::Sender<T>> for MustUseOneshotSender<T> {
-    fn from(sender: oneshot::Sender<T>) -> Self {
-        MustUseOneshotSender { tx: Some(sender) }
-    }
-}
-
-impl<T: std::fmt::Debug> Drop for MustUseOneshotSender<T> {
+impl Drop for MustUseClientResponseSender {
     #[instrument(skip(self))]
     fn drop(&mut self) {
         // we don't panic if we are shutting down anyway
         if !zebra_chain::shutdown::is_shutting_down() {
             // is_canceled() will not panic, because we check is_none() first
             assert!(
                 self.tx.is_none() || self.is_canceled(),
-                "unused oneshot sender: oneshot must be used or canceled: {:?}",
+                "unused client response sender: oneshot must be used or canceled: {:?}",
                 self
             );
         }
     }
 }
 
+impl MissingInventoryCollector {
+    /// Returns a newly created missing inventory collector, if needed.
+    ///
+    /// If `request` or the response contains missing inventory,
+    /// it is forwarded to the `inv_collector`, for the peer at `transient_addr`.
+    pub fn new(
+        request: &Request,
+        inv_collector: Option<broadcast::Sender<InventoryChange>>,
+        transient_addr: Option<SocketAddr>,
+    ) -> Option<Box<MissingInventoryCollector>> {
+        if !request.is_inventory_download() {
+            return None;
+        }
+
+        if let (Some(inv_collector), Some(transient_addr)) = (inv_collector, transient_addr) {
+            Some(Box::new(MissingInventoryCollector {
+                request: request.clone(),
+                collector: inv_collector,
+                transient_addr,
+            }))
+        } else {
+            None
+        }
+    }
+
+    /// Forwards any missing inventory to the registry.
+    ///
+    /// `zcashd` doesn't send `notfound` messages for blocks,
+    /// so we need to track missing blocks ourselves.
+    ///
+    /// This can sometimes send duplicate missing inventory,
+    /// but the registry ignores duplicates anyway.
+    pub fn send(self, response: &Result<Response, SharedPeerError>) {
```

Comment:
This new `send` method keeps track of peers that are missing blocks, even when they don't send us `notfound` messages.

`zcashd` only sends `notfound` for transactions - it just ignores requests for missing blocks with no response.